File: web/packages/excalidraw-svelte/src/model/value-types.ts
import type { BindMode } from "./enums.js";

/** A 2D point in `.excalidraw` wire form: a two-element `[x, y]` array. */
export type LocalPoint = [number, number];

/** Arbitrary JSON value, used for `customData` and unmodelled extras. */
export type JSONValue =
  | null
  | boolean
  | number
  | string
  | JSONValue[]
  | { [key: string]: JSONValue };

/** Corner rounding descriptor (`roundness: { type, value? } | null`). */
export interface Roundness {
  type: number;
  value?: number;
}

/** Reference to an element bound to this one (`BoundElement`). */
export interface BoundElement {
  id: string;
  type: "arrow" | "text";
}

/** Arrow-to-shape binding in the current `fixedPoint` + `mode` form. */
export interface FixedPointBinding {
  elementId: string;
  fixedPoint: LocalPoint;
  mode: BindMode;
}

/** A user-pinned segment of an elbow arrow (`FixedSegment`). */
export interface FixedSegment {
  start: LocalPoint;
  end: LocalPoint;
  index: number;
}

/** Image crop rectangle in natural-image coordinates (`ImageCrop`). */
export interface ImageCrop {
  x: number;
  y: number;
  width: number;
  height: number;
  naturalWidth: number;
  naturalHeight: number;
}

/** A persisted image/binary referenced by an image element (`BinaryFileData`). */
export interface BinaryFileData {
  mimeType: string;
  id: string;
  dataURL: string;
  created: number;
  lastRetrieved?: number;
  version?: number;
}

/**
 * Editor state carried in a `.excalidraw` file. Only a subset of the runtime
 * AppState is persisted and it varies by export source, so we keep the raw
 * key/value bag verbatim (lossless round-trip).
 */
export type AppState = Record<string, JSONValue>;

export function viewBackgroundColor(appState: AppState): string | undefined {
  const value = appState.viewBackgroundColor;
  return typeof value === "string" ? value : undefined;
}

export function gridModeEnabled(appState: AppState): boolean | undefined {
  const value = appState.gridModeEnabled;
  return typeof value === "boolean" ? value : undefined;
}

File: Sources/ExcalidrawModel/ValueTypes.swift
import ExcalidrawMath
import Foundation

/// Corner rounding descriptor (`roundness: { type, value? } | null`).
public struct Roundness: Codable, Equatable, Sendable {
    public var type: Int
    public var value: Double?

    public init(type: Int, value: Double? = nil) {
        self.type = type
        self.value = value
    }
}

/// Reference to an element bound to this one (`BoundElement`).
public struct BoundElement: Codable, Equatable, Sendable {
    public enum Kind: String, Codable, Sendable { case arrow, text }
    public var id: String
    public var type: Kind

    public init(id: String, type: Kind) {
        self.id = id
        self.type = type
    }
}

/// Arrow-to-shape binding in the current (`fixedPoint` + `mode`) form
/// (`FixedPointBinding`). Legacy `focus`/`gap` bindings are migrated on load by
/// the restore step (Phase 1, incremental).
public struct FixedPointBinding: Codable, Equatable, Sendable {
    public var elementId: String
    public var fixedPoint: Point
    public var mode: BindMode

    public init(elementId: String, fixedPoint: Point, mode: BindMode) {
        self.elementId = elementId
        self.fixedPoint = fixedPoint
        self.mode = mode
    }

    private enum CodingKeys: String, CodingKey { case elementId, fixedPoint, mode }

    /// Tolerant decode: real-world bindings (an agent-authored connector, or
    /// upstream Excalidraw's focus/gap binding) often carry only `elementId` and
    /// omit `fixedPoint`/`mode`. Those keys default here so one such binding can't
    /// make the whole `[ExcalidrawElement]` decode throw — which would otherwise
    /// blank the entire board on a strict (Swift) client.
    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        elementId = try c.decodeIfPresent(String.self, forKey: .elementId) ?? ""
        fixedPoint = try c.decodeIfPresent(Point.self, forKey: .fixedPoint) ?? Point(0, 0)
        mode = try c.decodeIfPresent(BindMode.self, forKey: .mode) ?? .orbit
    }
}

/// A user-pinned segment of an elbow arrow (`FixedSegment`).
public struct FixedSegment: Codable, Equatable, Sendable {
    public var start: Point
    public var end: Point
    public var index: Int

    public init(start: Point, end: Point, index: Int) {
        self.start = start
        self.end = end
        self.index = index
    }
}

/// Image crop rectangle in natural-image coordinates (`ImageCrop`).
public struct ImageCrop: Codable, Equatable, Sendable {
    public var x: Double
    public var y: Double
    public var width: Double
    public var height: Double
    public var naturalWidth: Double
    public var naturalHeight: Double

    public init(
        x: Double, y: Double, width: Double, height: Double,
        naturalWidth: Double, naturalHeight: Double
    ) {
        self.x = x
        self.y = y
        self.width = width
        self.height = height
        self.naturalWidth = naturalWidth
        self.naturalHeight = naturalHeight
    }
}

/// Arbitrary JSON value, used for `customData` and unmodelled extras so they
/// round-trip losslessly.
public enum JSONValue: Codable, Equatable, Sendable {
    case null
    case bool(Bool)
    case number(Double)
    case string(String)
    case array([JSONValue])
    case object([String: JSONValue])

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let value = try? container.decode(Bool.self) {
            self = .bool(value)
        } else if let value = try? container.decode(Double.self) {
            self = .number(value)
        } else if let value = try? container.decode(String.self) {
            self = .string(value)
        } else if let value = try? container.decode([JSONValue].self) {
            self = .array(value)
        } else {
            self = try .object(container.decode([String: JSONValue].self))
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .null: try container.encodeNil()
        case let .bool(value): try container.encode(value)
        case let .number(value): try container.encode(value)
        case let .string(value): try container.encode(value)
        case let .array(value): try container.encode(value)
        case let .object(value): try container.encode(value)
        }
    }
}
